core/integration/directory_test.go

```diff
@@ -300,3 +300,129 @@ func TestDirectoryWithCopiedFile(t *testing.T) {
 	require.NotEmpty(t, res.Directory.WithCopiedFile.File.ID)
 	require.Equal(t, "some-content", res.Directory.WithCopiedFile.File.Contents)
 }
+
+func TestDirectoryWithoutDirectory(t *testing.T) {
+	t.Parallel()
+
+	dirID := newDirWithFiles(t,
+		"some-file", "some-content",
+		"some-dir/sub-file", "sub-content")
+
+	var res2 struct {
+		Directory struct {
+			WithDirectory struct {
+				WithoutDirectory struct {
+					Contents []string
+				}
+			}
+		}
+	}
+
+	err := testutil.Query(
+		`query Test($src: DirectoryID!) {
+			directory {
+				withDirectory(path: "with-dir", directory: $src) {
+					withoutDirectory(path: "with-dir/some-dir") {
+						contents(path: "with-dir")
+					}
+				}
+			}
+		}`, &res2, &testutil.QueryOptions{
+			Variables: map[string]any{
+				"src": dirID,
+			},
+		})
+	require.NoError(t, err)
+	require.Equal(t, []string{"some-file"}, res2.Directory.WithDirectory.WithoutDirectory.Contents)
+}
+
+func TestDirectoryWithoutFile(t *testing.T) {
+	t.Parallel()
+
+	dirID := newDirWithFiles(t,
+		"some-file", "some-content",
+		"some-dir/sub-file", "sub-content")
+
+	var res2 struct {
+		Directory struct {
+			WithDirectory struct {
+				WithoutFile struct {
+					Contents []string
+				}
+			}
+		}
+	}
+
+	err := testutil.Query(
+		`query Test($src: DirectoryID!) {
+			directory {
+				withDirectory(path: "with-dir", directory: $src) {
+					withoutFile(path: "with-dir/some-file") {
+						contents(path: "with-dir")
+					}
+				}
+			}
+		}`, &res2, &testutil.QueryOptions{
+			Variables: map[string]any{
+				"src": dirID,
+			},
+		})
+	require.NoError(t, err)
+	require.Equal(t, []string{"some-dir"}, res2.Directory.WithDirectory.WithoutFile.Contents)
+}
+
+func TestDirectoryDiff(t *testing.T) {
+	t.Parallel()
+
+	aID := newDirWithFile(t, "a-file", "a-content")
+	bID := newDirWithFile(t, "b-file", "b-content")
+
+	var res struct {
+		Directory struct {
+			Diff struct {
+				Contents []string
+			}
+		}
+	}
+
+	diff := `query Diff($id: DirectoryID!, $other: DirectoryID!) {
+			directory(id: $id) {
+				diff(other: $other) {
+					contents
+				}
+			}
+		}`
+	err := testutil.Query(diff, &res, &testutil.QueryOptions{
+		Variables: map[string]any{
+			"id":    aID,
+			"other": bID,
+		},
+	})
+	require.NoError(t, err)
+
+	require.Equal(t, []string{"b-file"}, res.Directory.Diff.Contents)
+
+	err = testutil.Query(diff, &res, &testutil.QueryOptions{
+		Variables: map[string]any{
+			"id":    bID,
+			"other": aID,
+		},
+	})
+	require.NoError(t, err)
+
+	require.Equal(t, []string{"a-file"}, res.Directory.Diff.Contents)
+
+	/*
+		This triggers a nil panic in Buildkit!
```

Comment:
😮 Fortunately you have me to blame for that most likely. Sad though because we have a test case for "self-diff" in buildkit, so not sure what's different here. Do you have a stack trace on hand?

Reply:
:laughing: Yep, luckily I left that term sitting around:

```
time="2022-10-11T17:26:14Z" level=debug msg="load cache for diff (mkfile /a-file) -> (mkfile /a-file) with j3djj6hrknttubm7bfl24hp68::"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa8 pc=0xd4150a]
goroutine 89264 [running]:
github.com/moby/buildkit/frontend/gateway.(*llbBridgeForwarder).ReadDir(0xc00515e0b0, {0x154cc48, 0xc002d413e0}, 0xc005a20410)
	/src/frontend/gateway/gateway.go:827 +0x18a
github.com/moby/buildkit/control/gateway.(*GatewayForwarder).ReadDir(0x0, {0x154cc48, 0xc002d413e0}, 0x0)
	/src/control/gateway/gateway.go:145 +0x6f
github.com/moby/buildkit/frontend/gateway/pb._LLBBridge_ReadDir_Handler.func1({0x154cc48, 0xc002d413e0}, {0x1327c20, 0xc005a20410})
	/src/frontend/gateway/pb/gateway.pb.go:2701 +0x78
github.com/moby/buildkit/util/grpcerrors.UnaryServerInterceptor({0x154cc48, 0xc002d413e0}, {0x1327c20, 0xc005a20410}, 0x0, 0x0)
	/src/util/grpcerrors/intercept.go:14 +0x3d
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x154cc48, 0xc002d413e0}, {0x1327c20, 0xc005a20410})
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x3a
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1({0x154cba0, 0xc005107680}, {0x1327c20, 0xc005a20410}, 0xc000927700, 0xc000927720)
	/src/vendor/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:325 +0x61c
main.unaryInterceptor.func1({0x154cc48, 0xc002d41230}, {0x1327c20, 0xc005a20410}, 0xc000927700, 0xc000927720)
	/src/cmd/buildkitd/main.go:566 +0x1b8
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x154cc48, 0xc002d41230}, {0x1327c20, 0xc005a20410})
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x3a
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1({0x154cc48, 0xc002d41230}, {0x1327c20, 0xc005a20410}, 0xc00238abd0, 0x11c5480)
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34 +0xbf
github.com/moby/buildkit/frontend/gateway/pb._LLBBridge_ReadDir_Handler({0x131f000, 0xc000ca4060}, {0x154cc48, 0xc002d41230}, 0xc00465c9c0, 0xc0007a8b10)
	/src/frontend/gateway/pb/gateway.pb.go:2703 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00068ca80, {0x1565350, 0xc000cd8ea0}, 0xc0039ac360, 0xc000ca4390, 0x1e847e8, 0x0)
	/src/vendor/google.golang.org/grpc/server.go:1282 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc00068ca80, {0x1565350, 0xc000cd8ea0}, 0xc0039ac360, 0x0)
	/src/vendor/google.golang.org/grpc/server.go:1619 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/src/vendor/google.golang.org/grpc/server.go:921 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/src/vendor/google.golang.org/grpc/server.go:919 +0x294
```

Which points here:

https://github.com/moby/buildkit/blob/v0.10.4/frontend/gateway/gateway.go#L827

Reply:
Oh okay, there's actually an opportunity for me to possibly pass the blame then! 😄

There were apparently some fixes around nils and refs recently made: https://github.com/moby/buildkit/issues/3134#issuecomment-1275460396

The exception is happening on a much higher level than the actual diff implementation, which is why there's a chance it's something more related to refs. I'd want to check on the latest buildkit builds before reporting this upstream. However, that's not a high priority at all for this PR, maybe just include a comment here saying that we can try uncommenting when we upgrade buildkit (or make an issue for it).

Reply:
> Oh okay, there's actually an opportunity for me to possibly pass the blame then! 😄

Should have known I was dooming myself as soon as I said this. I just looked and realized it's still present in Buildkit master. The previous test suite for diffop missed it because it can only happen when you use the gateway api (not solve directly). 

Sent out a PR to fix here: https://github.com/moby/buildkit/pull/3169